ios/sound.mm

```diff
@@ -24,209 +24,275 @@
-        lNumInChanPlusAddChan = 8;
-
-        for ( int iCh = 0; iCh < lNumInChanPlusAddChan; iCh++ )
-        {
-            int iSelCH, iSelAddCH;
+    QMutexLocker locker ( &pSound->MutexAudioProcessCallback );
+    Float32*     pData = static_cast<Float32*> ( inInputData->mBuffers[0].mData );
 
-            GetSelCHAndAddCH ( iCh, lNumInChan, iSelCH, iSelAddCH );
-
-            if ( iSelAddCH >= 0 )
-            {
-                // for mixed channels, show both audio channel names to be mixed
-                channelInputName[iCh] =
-                    channelInputName[iSelCH] + " + " + channelInputName[iSelAddCH];
-            }
-        }
-    }
-    else
+    // copy input data
+    for ( int i = 0; i < pSound->iCoreAudioBufferSizeMono; i++ )
     {
-        // regular case: no mixing input channels used
-        lNumInChanPlusAddChan = lNumInChan;
+        // copy left and right channels separately
+        pSound->vecsTmpAudioSndCrdStereo[2 * i]     = (short) ( pData[2 * i] * _MAXSHORT );     // left
+        pSound->vecsTmpAudioSndCrdStereo[2 * i + 1] = (short) ( pData[2 * i + 1] * _MAXSHORT ); // right
     }
-*/
-    // everything is ok, return empty string for "no error" case
-    return "";
+    pSound->ProcessCallback ( pSound->vecsTmpAudioSndCrdStereo );
 }
 
-void CSound::SetLeftInputChannel  ( const int iNewChan )
+// TODO - CSound::Init is called multiple times at launch to verify device capabilities.
+// For iOS though, Init takes long, so should better reduce those inits for iOS (Now: 9 times/launch).
+int CSound::Init ( const int iCoreAudioBufferSizeMono )
 {
-    /*// apply parameter after input parameter check
-    if ( ( iNewChan >= 0 ) && ( iNewChan < iNumInChanPlusAddChan ) )
+    try
     {
-        iSelInputLeftChannel = iNewChan;
+        this->iCoreAudioBufferSizeMono = iCoreAudioBufferSizeMono;
+
+        // set internal buffer size value and calculate stereo buffer size
+        iCoreAudioBufferSizeStereo = 2 * iCoreAudioBufferSizeMono;
+
+        // create memory for intermediate audio buffer
+        vecsTmpAudioSndCrdStereo.Init ( iCoreAudioBufferSizeStereo );
+
+        AVAudioSession* sessionInstance = [AVAudioSession sharedInstance];
+
+        // we are going to play and record so we pick that category
+        NSError* error = nil;
+        [sessionInstance setCategory:AVAudioSessionCategoryPlayAndRecord
+                         withOptions:AVAudioSessionCategoryOptionMixWithOthers | AVAudioSessionCategoryOptionDefaultToSpeaker |
+                                     AVAudioSessionCategoryOptionAllowBluetooth | AVAudioSessionCategoryOptionAllowBluetoothA2DP |
+                                     AVAudioSessionCategoryOptionAllowAirPlay
+                               error:&error];
+
+        // using values from jamulus settings 64 = 2.67ms/2
+        NSTimeInterval bufferDuration = iCoreAudioBufferSizeMono / 48000.0; // yeah it's math
```

Comment:
This is not defined somewhere already? 48k magic number?

Reply:
If it is the sample rate, there should be a constant somewhere. SYSTEM_SAMPLE_RATE or PREFERRED_SAMPLE_RATE need to check the Code.

Reply:
updated!